src/Microsoft.AspNetCore.OData/Formatter/ODataInputFormatter.cs

```diff
@@ -151,25 +143,27 @@ public override Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterCo
 
                 ODataDeserializerProvider deserializerProvider = request.GetRequestContainer().GetRequiredService<ODataDeserializerProvider>();
 
-                object result = ODataInputFormatterHelper.ReadFromStream(
+                Task<object> resultTask = ODataInputFormatterHelper.ReadFromStreamAsync(
                     type,
                     defaultValue,
                     request.GetModel(),
                     GetBaseAddressInternal(request),
                     new WebApiRequestMessage(request),
-                    () => ODataMessageWrapperHelper.Create(request.Body, request.Headers, request.GetODataContentIdMapping(), request.GetRequestContainer()),
+                    () => ODataMessageWrapperHelper.Create(new StreamWrapper(request.Body), request.Headers, request.GetODataContentIdMapping(), request.GetRequestContainer()),
                     (objectType) => deserializerProvider.GetEdmTypeDeserializer(objectType),
                     (objectType) => deserializerProvider.GetODataDeserializer(objectType, request),
                     getODataDeserializerContext,
                     (disposable) => toDispose.Add(disposable),
                     logErrorAction);
 
+                resultTask.Wait();
```

Comment:
Don't Wait here, just await above.

Reply:
Originally I just did await above (along with the async modifier on the method), but that caused some of the existing tests to fail (I think primarily in the error cases) so I had to do it this way for backward compatibility (since this is unfortunately a public API I couldn't just fix the tests.)

---
In reply to: [515363752](https://github.com/OData/WebApi/pull/2336#discussion_r515363752) [](ancestors = 515363752)

Reply:
I'm not sure what tests broke but doing this wait basically negates the entire fix.

Reply:
Got it. I'll put it back and disable the test for now -- it was verifying an expected exception from the underlying method, and getting back an aggregation exception instead of the expected argument null exception. that difference won't affect the test release, but for the final release I may have to create two versions of the method to avoid a possible breaking change.

---
In reply to: [515383405](https://github.com/OData/WebApi/pull/2336#discussion_r515383405) [](ancestors = 515383405)

Reply:
What besides a test would be able to observe this breaking change? The formatter is called invoked via MVC.

Reply:
There would be a different exception surfaced to the developer -- instead of getting a meaningful exception they would get a generic "AggregateException"

---
In reply to: [515997243](https://github.com/OData/WebApi/pull/2336#discussion_r515997243) [](ancestors = 515997243)